File: common/src/main/java/juuxel/adorn/util/PropertyRef.java
package juuxel.adorn.util;

import java.lang.invoke.MethodHandles;
import java.util.Map;

/**
 * Reference to a mutable property.
 */
public interface PropertyRef<T> {
    String getName();
    T get();
    void set(T value);

    static <T> PropertyRef<T> ofField(Object owner, String fieldName) {
        try {
            var field = owner.getClass().getField(fieldName);
            field.setAccessible(true);

            var name = field.getName();
            var lookup = MethodHandles.lookup();
            var getter = lookup.unreflectGetter(field);
            var setter = lookup.unreflectSetter(field);

            return new PropertyRef<>() {
                @Override
                public String getName() {
                    return name;
                }

                @SuppressWarnings("unchecked")
                @Override
                public T get() {
                    try {
                        return (T) getter.invoke(owner);
                    } catch (Throwable e) {
                        throw new RuntimeException(e);
                    }
                }

                @Override
                public void set(T value) {
                    try {
                        setter.invoke(owner, value);
                    } catch (Throwable e) {
                        throw new RuntimeException(e);
                    }
                }
            };
        } catch (ReflectiveOperationException e) {
            throw new RuntimeException(e);
        }
    }

    static <T> PropertyRef<T> ofMapEntry(Map<String, T> map, String key) {
        return new PropertyRef<>() {
            @Override
            public String getName() {
                return key;
            }

            @Override
            public T get() {
                return map.get(key);
            }

            @Override
            public void set(T value) {
                map.put(key, value);
            }
        };
    }
}

File: common/src/main/kotlin/juuxel/adorn/util/PropertyRef.kt
package juuxel.adorn.util

import java.lang.invoke.MethodHandles
import kotlin.reflect.KProperty

/**
 * Reference to a mutable property.
 */
interface PropertyRef<T> {
    val name: String
    fun get(): T
    fun set(value: T)

    companion object {
        /**
         * Creates a reflected property reference for a field of the [owner].
         */
        fun <T> ofField(owner: Any, fieldName: String): PropertyRef<T> {
            val field = owner::class.java.getField(fieldName)
            field.isAccessible = true

            val name = field.name
            val lookup = MethodHandles.lookup()
            val getter = lookup.unreflectGetter(field)
            val setter = lookup.unreflectSetter(field)

            return object : PropertyRef<T> {
                override val name = name

                @Suppress("UNCHECKED_CAST")
                override fun get(): T =
                    getter.invoke(owner) as T

                override fun set(value: T) {
                    setter.invoke(owner, value)
                }
            }
        }
    }
}

/**
 * Creates a reflected property reference for a [JvmField] property of `this` object.
 */
inline fun <O : Any, T> O.ref(finder: (O) -> KProperty<T>): PropertyRef<T> {
    return PropertyRef.ofField(this, finder(this).name)
}
